ambry-commons/src/main/java/com.github.ambry.commons/BlobId.java

```diff
@@ -24,28 +25,116 @@
+ *   blob is assigned.
+ * </p>
+ * <pre>
+ * +---------------------------------------------+
+ * | version | partitionId | uuidSize | uuid     |
+ * | (short) | (n bytes)   | (short)  | (n bytes)|
+ * +---------------------------------------------+
+ * </pre>
+ * <p>
+ *   Version 2, which includes{@code flag}, {@code accountId}, {@code containerId}, {@code datacenterId}, and
+ *   {@code partitionId} of the blob. {@code flag} is a single byte that carries the meta information for this
+ *   blobId. The {@code datacenterId} is the id of the datacenter where this blob was originally posted (not
+ *   through replication). The {@code accountId} is the id of the {@link Account} the blob belongs to. The
+ *   {@code containerId} is the {@link Container} the blob belongs to. The {@code partitionId} is the
+ *   {@link com.github.ambry.clustermap.Partition} to which this blob is assigned.
+ * </p>
+ * <pre>
+ * +---------+-------+--------------+-----------+-------------+-------------+----------+----------+
+ * | version | flag  | datacenterId | accountId | containerId | partitionId | uuidSize | uuid     |
+ * | (short) | (byte)| (byte)       | (short)   | (short)     | (n bytes)   | (short)  | (n bytes)|
+ * +---------+----------------------+-----------+-------------+-------------+----------+----------+
+ * </pre>
  */
 public class BlobId extends StoreKey {
-  private Short version = 1;
-  private static short Version_Size_In_Bytes = 2;
+  static final short BLOB_ID_V1 = 1;
+  static final short BLOB_ID_V2 = 2;
+  static final byte DEFAULT_FLAG = 0;
+  private static final short VERSION_SIZE_IN_BYTES = 2;
+  private static final int UUID_SIZE_IN_BYTES = 4;
+  private static final short FLAG_SIZE_IN_BYTES = 1;
+  private static final short DATACENTER_ID_SIZE_IN_BYTES = 1;
+  private static final short ACCOUNT_ID_SIZE_IN_BYTES = 2;
+  private static final short CONTAINER_ID_SIZE_IN_BYTES = 2;
+  // the version to indicate the serialized format.
+  private Short version;
+  private Byte flag;
+  private Byte datacenterId;
+  private Short accountId;
+  private Short containerId;
   private PartitionId partitionId;
   private String uuid;
-  private static int UUID_Size_In_Bytes = 4;
 
   /**
-   * Constructs a new unique BlobId for the specified partition.
+   * Constructs a new unique BlobId. This will construct a blob in version 2, so the serialization embeds accountId,
+   * containerId, datacenterId, and partitionId.
+   * @param flag A byte to embed additional information of this blobId.
+   * @param datacenterId The id of the datacenter to be embedded into the blob.
+   * @param accountId The id of the {@link Account} to be embedded into the blob.
+   * @param containerId The id of the {@link Container} to be embedded into the blob.
+   * @param partitionId The partition where this blob is to be stored.
+   */
+  BlobId(Byte flag, Byte datacenterId, Short accountId, Short containerId, PartitionId partitionId) {
```

Comment:
It is called by the `Builder`, and now we have the style that `Builder` and the actual component are in different classes.